frontend/src/lib/components/modal/TimedModal.svelte

```diff
@@ -0,0 +1,113 @@
+<script lang="ts">
+  import {tweened} from 'svelte/motion';
+  import {onDestroy, onMount} from 'svelte';
+
+  export let onTimeout: () => void; // function to be triggered after time runs out
+  export let timerDuration = 30; // logout timer duration in seconds
+  export let timeLeftInt: number = Math.ceil(timerDuration); // time left in seconds (int)
+
+  let progressBarTimer = tweened(timerDuration, {duration: 0}); // used for progress bar animation
+  let isOpen = false; // variable for the modal state
+  let modalContainer: HTMLDialogElement | null = null; // container element for the modal
+  let timeout: NodeJS.Timeout | null = null; // timeout for triggering onTimeout()
+
+  export const openModal = () => {
+    if (!isOpen) {
+      modalContainer?.showModal();
+      isOpen = true;
+      resetProgressBarTimer();
+      resetTimeout();
+    }
+  };
+  export const closeModal = () => {
+    if (isOpen) {
+      modalContainer?.close();
+      isOpen = false;
+      stopProgressBarTimer();
+    }
+  };
+
+  // update progressBarTimerInt to an integer when progressBarTimer updates
+  $: {
+    if ($progressBarTimer) {
+      const t = Math.ceil($progressBarTimer);
+      if (t < timeLeftInt) {
+        timeLeftInt = t;
+      }
+    }
+  }
+
+  // function triggered by pressing the escape button
+  const handleEscape = (e: KeyboardEvent) => {
+    if (isOpen && e.key == 'Escape') {
+      closeModal();
+    }
+  };
+
+  // add and remove event handles on mount and unmount
+  onMount(() => {
+    document.addEventListener('keydown', handleEscape);
+  });
+  onDestroy(() => {
+    document.removeEventListener('keydown', handleEscape);
+  });
+
+  const resetTimeout = () => {
+    if (timeout) clearTimeout(timeout);
+    timeout = setTimeout(() => {
+      if (isOpen) {
+        closeModal();
+        onTimeout();
+        stopProgressBarTimer();
+      }
+    }, timerDuration * 1000);
+  };
+
+  const resetProgressBarTimer = () => {
+    progressBarTimer = tweened(timerDuration, {
+      duration: timerDuration * 1000
+    });
+    timeLeftInt = timerDuration;
+    $progressBarTimer = 0;
+  };
+
+  const stopProgressBarTimer = () => {
+    progressBarTimer.set($progressBarTimer, {duration: 0});
+    if (timeout) clearTimeout(timeout);
+    timeout = null;
+  };
+</script>
+
+<!--
+@component
+TimedModal is a modal that will automatically close after a set amount of time.
+
```

Comment:
I think the section about keyboard navigation is redundant because only developers read the JSDoc and they don't really care about the keyboard navigation functionality. I'll still add it.

Reply:
Thanks! I think it's good to include the kbd nav description in this and other components, because it's part of the component's functionality and because accessibility is a special concern of ours.